Frontend: Cleanup Monitor() function

uablrek opened the pull request:
## Description

There seems to be a bug in here somewhere. But I couldn't follow all logic so I made a general cleanup.

One thing that might have been a bug is the "break" in
https://github.com/Nordix/Meridio/blob/60a45b5a8254ada9843466871abccc62eb9f1f30/cmd/frontend/internal/frontend/service.go#L314

The Monitor can exit more or less silently and announceFe will not be called in intervals ever after.

## Issue link

May fix the NSP-FE issue described in https://github.com/Nordix/Meridio/issues/355#issuecomment-1514477574

## Checklist

- Purpose
    - [x] Bug fix
    - [ ] New functionality
    - [ ] Documentation
    - [x] Refactoring
    - [ ] CI
- Test
    - [ ] Unit test
    - [ ] E2E Test
    - [x] Tested manually
- Introduce a breaking change
    - [ ] Yes (description required)
    - [x] No

uablrek commented:
I added timeouts of 5s (hard-coded) for communication with the NSP

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -288,95 +293,79 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+					if err := announceFrontend(ctx, fes.targetRegistryClient); err != nil {
+						logger.Error(err, "Announce frontend connectivity")
 					}
+					// refresh NSP entry (even if announceFrontend() above fails?)
```
Yes, even if the send towards NSP fails, the FE still has external connectivity. The FE should keep trying to inform NSP about it.

zolug replied:
That question is still in place. Would be better to remove, in order to avoid confusion.
Without the retry block, the FE would not be able to announce its connectivity without a prior state change (-> down, up).

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -288,95 +293,79 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+				// although configured at least one IP family has no connectivity
+				if hasConnectivity {
+					denounce = true
+					// should this be moved to "if denounce" above?
```
It could be moved there, yes. It reflects the status of external connectivity and is used by probes.

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -258,25 +258,30 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+		hasConnectivity := false
 		connectivityMap := map[string]bool{}
 		delay := 3 * time.Second // when started grant more time to write the whole config (minimize intial link flapping)
 		_ = fes.WaitStart(ctx)
```
The following code from the start of Monitor() should be moved after WaitStart() as well. The error _should not happen_ that case.
```
lp, err := fes.routingService.LookupCli()
	if err != nil {
		errCh <- fmt.Errorf("routing service cli not found: %v", err)
		return
	}
```

uablrek replied:
Can WaitStart() be moved to before the go func()?

uablrek replied:
You are right. I move them back into the go function.

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -288,95 +293,79 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+			bfdOut, err := fes.routingService.ShowBfdSessions(ctx, lp, `NBR-BFD`)
+			if err != nil {
+				logger.Error(err, "BFD output", "out", strings.Split(bfdOut, "\n"))
+				denounce = true
```
Again, I don't think we should do denounce.

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -258,25 +258,30 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+					logger.Error(err, "denounceFrontend")
+				}
+				connectivityMap = map[string]bool{}
+				delay = 3 * time.Second // Avoid flapping (more?)
```
Not sure about this one. An env var would be nice.

uablrek replied:
ok

uablrek replied:
Honestly I have no idea what those errors I called "session errors" are and what that can cause them. So, please take a close look, and suggest what to do differently if needed

uablrek commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -258,25 +258,30 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+
 			select {
 			case <-ctx.Done():
 				logger.Info("Shutting down")
```
Ok, I add that. But with a brand new context

uablrek commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -288,95 +293,79 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
-					logger.Info("protocol output", "out", protocolOut)
-					//linkCh <- "No protocols match"
+				logger.Error(err, "protocol output", "out", strings.Split(protocolOut, "\n"))
+				denounce = true
```
I will add a counter for the "session errors" and denounce after X tries (X configurable)

zolug replied:
Btw, whenever the code deems denounce is required, the VIPs must be denounced as well. (Unless the FE is terminating/restarting).
Otherwise, for example if the connection is still UP according to bird, then BGP will still attract traffic.

Similarly, health.SetServingStatus should be adjusted.

uablrek commented:
Updated. I will merge and force-push after all is settled. I have not tested to set the env variables. I wait with that until I have fixed all review comments.
Also the announce log is temporary on DEBUG, I must remember to set it back to TRACE.
BTW is the 5s hard-coded timeout for announce/denounce ok?

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -293,25 +335,25 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
 			if strings.Contains(protocolOut, bird.NoProtocolsLog) {
 				logger.Info("protocol output", "out", protocolOut)
-				denounce = true
+				sessionErrors++
```
This is not a "session error". It indicates that there are no gateways configured. Either not added yet (in which case it doesn't matter if it was a session error or not), or all the gateways had been removed from the Attractor (FE config).
A denounce is required.

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -252,24 +252,66 @@ func (fes *FrontEndService) SetNewConfig(ctx context.Context, c interface{}) err
+		// sessionErrors may be temporary
+		sessionErrors int
+		// maxSessionErrors is max sessionErrors before a denounce
+		maxSessionErrors int = 3
```
No clue why these errors pop up. But I would prefer a bit higher value, maybe 5.

zolug commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -330,6 +333,12 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
 				}
+				if hasConnectivity {
+					// (logging will not work inside denounceFrontend())
+					if err := denounceFrontend(context.Background(), fes.targetRegistryClient); err != nil {
```
I wonder if blocking the shutdown for 5 seconds could be a bit too much if e.g. NSP is not around.
Anyways, we can revisit it, if required...

uablrek commented on cmd/frontend/internal/frontend/service.go:
```diff
@@ -258,25 +258,30 @@ func (fes *FrontEndService) Monitor(ctx context.Context, errCh chan<- error) {
+					refreshCancel()
+					refreshCancel = nil
+				}
+				if err := denounceFrontend(ctx, fes.targetRegistryClient); err != nil {
```
> or some simple but lame init delay could be used.

I go for this option as a start. Same as before, a 3s delay

zolug replied:
Seems that in the baseline it wasn't the delay rather the grpc option WaitForReady that alowed the initial (or any) denounce to succeed. It blocks RPCs until the connection is ready.

Even with the 3 seconds delay, denounce fails on xcluster for me.

I think the easiest way forward would be if denounce did not have hard coded timeout, instead it would be the callers responsibility to pass a feasible context. Or maybe just increase the hard-coded value in denounce to 30 seconds, and in case of shutdown pass a context with lower timeout.

uablrek replied:
I think it may be best just to remove the timeout in announceFrontend/denounceFrontend entirely and make a comment. Not adding a ctx with timeout at all in the calls. I.e. same as it was before. Timeouts would be hard to decide, especially if they are dynamic. The passed ctx can still be canceled from main.

Or do you think there is a case where the calls can hang for an unreasonable time (whatever that is)?

uablrek replied:
> and in case of shutdown pass a context with lower timeout.

I missed this. Of course, a timeout is needed on shutdown

zolug replied:
You're right. Expect for the shutdown, I don't see much need for timeout.

uablrek commented:
I added a pointer to the config struct in `FrontEndService struct`. IMO that's better than to copy all config vars. Not only unnecessary code is removed, but also that the value is actually a configurable becomes clear.
